source/particles/particle_handler.cc

```diff
@@ -0,0 +1,1152 @@
+                    }
+                }
+              catch (typename Mapping<dim>::ExcTransformationFailed &)
+                {}
+            }
+
+          if (!found_cell)
+            {
+              // The particle is not in a neighbor of the old cell.
+              // Look for the new cell in the whole local domain.
+              // This case is rare.
+              try
+                {
+                  const std::pair<const typename Triangulation<dim,spacedim>::active_cell_iterator,
+                        Point<dim> > current_cell_and_position =
+                          GridTools::find_active_cell_around_point<> (*mapping,
+                                                                      *triangulation,
+                                                                      (*it)->get_location());
+                  current_cell = current_cell_and_position.first;
+                  current_reference_position = current_cell_and_position.second;
+                }
+              catch (GridTools::ExcPointNotFound<spacedim> &)
+                {
+                  // We can find no cell for this particle. It has left the
+                  // domain due to an integration error or an open boundary.
+                  continue;
+                }
+            }
+
+          // If we are here, we found a cell and reference position for this particle
+          (*it)->set_reference_location(current_reference_position);
+
+          // Reinsert the particle into our domain if we own its cell.
+          // Mark it for MPI transfer otherwise
+          if (current_cell->is_locally_owned())
+            {
+              sorted_particles.push_back(std::make_pair(internal::LevelInd(current_cell->level(),current_cell->index()),
+                                                        (*it)->particle->second));
+            }
+          else
+            {
+              moved_particles[current_cell->subdomain_id()].push_back(*it);
+              moved_cells[current_cell->subdomain_id()].push_back(current_cell);
+            }
+        }
+    }
+
+    // Sort the updated particles. This pre-sort speeds up inserting
+    // them into particles to O(N) complexity.
+    std::multimap<internal::LevelInd,Particle <dim,spacedim> > sorted_particles_map;
+
+    // Exchange particles between processors if we have more than one process
+#ifdef DEAL_II_WITH_MPI
+    if (dealii::Utilities::MPI::n_mpi_processes(triangulation->get_communicator()) > 1)
+      send_recv_particles(moved_particles,sorted_particles_map,moved_cells);
+#endif
+
+    sorted_particles_map.insert(sorted_particles.begin(),sorted_particles.end());
+
+    for (unsigned int i=0; i<particles_out_of_cell.size(); ++i)
+      remove_particle(particles_out_of_cell[i]);
+
+    particles.insert(sorted_particles_map.begin(),sorted_particles_map.end());
+    update_cached_numbers();
+  }
+
+
+
+#ifdef DEAL_II_WITH_MPI
+  template <int dim, int spacedim>
+  void
```

Comment:
This is the place where you'd move the `#ifdef` down into the body of the function.